wrap/unwrap transformations

davidercruz opened the pull request:
Partly fixes #1185 

Changes made in this Pull Request:
 -  changed make whole to work with triclinic boxes.

It's currently in the transformations submodule due to requiring apply_PBC from lib.distances. If I import it from lib.mdamath I get some errors - but that's probably me being a noob. 
I think it's faster than the old method, but I haven't done any exhaustive benchmarks. It takes 0.3 seconds for the adk_oplsaa system in the test datafiles, in my system.

Huge thanks to @mnmelo for helping me.

I will do the tests ASAP, and add a transformations that unwraps the system using make_whole. 

PR Checklist
------------
 - [ ] Tests?
 - [ ] Docs?
 - [ ] CHANGELOG updated?
 - [ ] Issue raised/referenced?

richardjgowers commented:
@davidercruz the problem you're having with putting it in `mdamath` is that at the point that `mdamath.py` is defined `distances.py` doesn't exist yet, you can see the order things are defined in the `__init__` of a module:

https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/lib/__init__.py

It looks like distance depends on mdamath, so you've got a circular dependency.  To get around this, import apply_PBC inside the function scope.

I think this needs some comments to explain how it works :)

jbarnoud commented:
It would be better to have `make_whole` in `lib.mdamath` for the moment. It does not really belongs there as it depends on the atom group, but having the new `make_whole` in the place of the old one means it is easier to review because we actually have a diff, and it benefits from all the tests that cover the old `make_whole`. Ultimately, the old one must disappear, we do ot want to have two functions with the same name and doing the same thing.

The function does not belong in the transformation module either by the way. It is not the transformation of a time step, but a more general utility function.

At some point, we should probably move `make_whole` somewhere more suitable. But this means that the version in `mdamath` must become an alias for the new one with a deprecation warning. I'd rather have that in a different PR to keep the focus on the feature. You can solve the circular dependency by importing `apply_PBC` in the function for the moment.

I would be curious to see a benchmark. It does not have to be an extensive one, though.

davidercruz commented:
@jbarnoud I replaced old make_whole with the new. I had to remove some tests that did not make sense with the new algorithm. I also improved the special check with the long comment that switches the order of the calculation arround. I found that in one of my trajectories repeats can happen without causing the loop to get stuck. However the order of the code was causing it to fail as the `newvecs` array was of a different shape of that of the positions. Now it's working fine.

@richardjgowers I added some comments explaining what happens.

One interesting thing about this new make_whole function is that it can work with whole atomgroup at once, instead of having to do it a single fragment at a time. I haven't tried it yet, but there's no reason for it not to work.

richardjgowers commented on package/MDAnalysis/lib/mdamath.py:
```diff
@@ -279,7 +279,7 @@ def _is_contiguous(atomgroup, atom):
     return True
 
 
-def make_whole(atomgroup, reference_atom=None):
```
So you've changed how this function can be called, which isn't usually allowed, ie maybe someone was using the `reference_atom` functionality and we can't take that away.

jbarnoud commented:
@richardgowers does your version handle triclinic boxes?

richardjgowers commented:
@jbarnoud yeah it will do once it's merged, I've just got to figure out how to link to our pbc code in calc_distances :)

davidercruz commented:
@richardjgowers In what cases do you think it will fail?

richardjgowers commented on package/MDAnalysis/lib/mdamath.py:
```diff
@@ -341,92 +329,71 @@ def make_whole(atomgroup, reference_atom=None):
+        # of the atom2 group, suchs cases lead to the while loop being stuck:
+        # this atom is moved back and forth between the two atoms it is conected
+        # to. Since were tracking the atoms that need changing, if this group stays
+        # the same over multiple iterations, we change atom1.positions instead.
```
@davidercruz If you had a ring molecule which had oddly/pathologically numbered bonds, I think you could end up chasing your tail because you're updating all positions at once.  This would lead to atoms from the ring jumping between each other images.

jbarnoud commented:
@davidercruz I need to read both PR carefully and to understand how they both work. You should have a look at #1965 and give it a review: you worked on the issue so you have ideas of what it involves. In any case, your input would be valuable.

You can also try your PR with a nightmare scenario and see if it get stuck. Something like a C60 fullerene may be nightmarish enough.

Finally, you have a test failure that indicates a small error in your code. You should fix it and see if the test pass. It is difficult to compare two implementations while not being sure about the validity of their outputs.

After these two last point, which should not be too long, you can put the PR on hold until I look at both codes. Hopefully, I'll have time tomorrow for that and your two other PRs. In the meantine, you can look at the per-axis/per-plane centering; or at the the fitting.

davidercruz commented:
@jbarnoud @richardjgowers 
Ok I tested it with fullerene and indeed it fails. While it doesn't enter an endless loop, it just doesn't unwrapps the molecule completely.

richardjgowers commented:
Can you post the fullerene so I can try with mine?
On Mon, Jul 2, 2018 at 12:14 PM, Davide Cruz <notifications@github.com>
wrote:

> @jbarnoud <https://github.com/jbarnoud> @richardjgowers
> <https://github.com/richardjgowers>
> Ok I tested it with fullerene and indeed it fails. While it doesn't enter
> an endless loop, it just doesn't unwrapps the molecule completely.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MDAnalysis/mdanalysis/pull/1961#issuecomment-401873230>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AI0jBym2ra6cCBIOOvGMIaVCeiOBncpoks5uClT0gaJpZM4U76XR>
> .
>

davidercruz commented:
@richardjgowers here goes
[fullerene.zip](https://github.com/MDAnalysis/mdanalysis/files/2156173/fullerene.zip)

richardjgowers commented:
@davidercruz it was a good approach, definitely vectorising the position updates will usually be faster.  Do you want to undo the make_whole changes and keep the unwrap transformation here?

davidercruz commented:
Let's undo it for now, and I will focus on other transformations. But I
would like to try this using graphs later.

On Mon, Jul 2, 2018 at 10:52 PM Richard Gowers <notifications@github.com>
wrote:

> @davidercruz <https://github.com/davidercruz> it was a good approach,
> definitely vectorising the position updates will usually be faster. Do you
> want to undo the make_whole changes and keep the unwrap transformation here?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MDAnalysis/mdanalysis/pull/1961#issuecomment-401949307>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AkAGGSe5Gks8squa1P4Qqx6Smd2uu8WRks5uCpYhgaJpZM4U76XR>
> .
>
